### PR TITLE
chore: [Multi-domain] Validate domain name

### PR DESCRIPTION
**Pull request by mjhenkes**
- Closes 

### User facing changelog

n/a

### Additional details

This pr uses our [existing](https://github.com/cypress-io/cypress/blob/develop/packages/driver/package.json#L55) domain validation library, [is-valid-domain](https://www.npmjs.com/package/is-valid-domain), to validate that the domain string is a valid domain, eg `foobar.com` and not a sub domain `eu.foobar.com`. As such I've slightly modified the error string.

### How has the user experience changed?

Screen Shot 2022-02-22 at 1 28 49 PM

### PR Tasks

- [x] Have tests been added/updated?
- [ ] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [ ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [ ] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
- [ ] Have new configuration options been added to the [`cypress.schema.json`](https://github.com/cypress-io/cypress/blob/develop/cli/schema/cypress.schema.json)?

### CODE REVIEW

File: packages/driver/cypress/integration/e2e/multi-domain/multi_domain_spec.ts [hunk 1]
```diff
@@ -171,6 +171,28 @@ describe('multi-domain', { experimentalSessionSupport: true, experimentalMultiDo
...
 
+    it('fails on a bad domain name', (done) => {
+      cy.on('fail', (err) => {
+        expect(err.message).to.equal('`cy.switchToDomain()` requires the first argument to be a string representing a domain that does not not include any sub domains. You passed: `foobar`')
+
+        done()
+      })
+
+      cy.switchToDomain('foobar', () => {
+      })
```
Comment (AtofStryker) [inline packages/driver/cypress/integration/e2e/multi-domain/multi_domain_spec.ts:182]: more of a nit here, but figure we don't need the function body?
```suggestion
      cy.switchToDomain('foobar', () => undefined)
```

File: packages/driver/src/cy/multi-domain/validator.ts [hunk 2]
```diff
@@ -12,7 +13,7 @@ export class Validator {
   }
 
   validate ({ callbackFn, data, domain, done, doneReference }) {
-    if (typeof domain !== 'string') {
+    if (typeof domain !== 'string' || !(isValidDomain(domain, { allowUnicode: true, subdomain: false }))) {
```
Comment (AtofStryker) [inline packages/driver/src/cy/multi-domain/validator.ts:16]: I was going to suggest maybe using `_.isString` here to account for the string constructor, like `new String('foobar.com')`, but turns out `is-valid-domain` doesn't account for that and actually does the check the string conditional we are doing [here](https://github.com/miguelmota/is-valid-domain/blob/master/index.js#L6). We might be able to remove the first conditional.
```suggestion
    if !(isValidDomain(domain, { allowUnicode: true, subdomain: false })) {
```

File: packages/driver/cypress/integration/e2e/multi-domain/multi_domain_spec.ts [hunk 3]
```diff
@@ -171,6 +171,28 @@ describe('multi-domain', { experimentalSessionSupport: true, experimentalMultiDo
...
+
+    it('fails on a domain name with a sub domain', (done) => {
+      cy.on('fail', (err) => {
+        expect(err.message).to.equal('`cy.switchToDomain()` requires the first argument to be a string representing a domain that does not not include any sub domains. You passed: `eu.foobar.com`')
+
+        done()
+      })
+
+      cy.switchToDomain('eu.foobar.com', () => {
+      })
```
Comment (AtofStryker) [inline packages/driver/cypress/integration/e2e/multi-domain/multi_domain_spec.ts:193]:
```suggestion
      cy.switchToDomain('eu.foobar.com', () => undefined)
```

**Review by AtofStryker — APPROVED** [review]

File: packages/driver/src/cypress/error_messages.ts [hunk 4]
```diff
@@ -1694,7 +1694,7 @@ export default {
       message: `${cmd('switchToDomain')} requires enabling the experimentalMultiDomain flag`,
     },
     invalid_domain_argument: {
-      message: `${cmd('switchToDomain')} requires the first argument to be a string. You passed: \`{{arg}}\``,
+      message: `${cmd('switchToDomain')} requires the first argument to be a string representing a domain that does not not include any sub domains. You passed: \`{{arg}}\``,
```
Comment (mschile) [inline packages/driver/src/cypress/error_messages.ts:1697]:
```suggestion
      message: `${cmd('switchToDomain')} requires the first argument to be a string representing a domain that does not not include any subdomains. You passed: \`{{arg}}\``,
```

**Review by mschile — APPROVED** [review]

File: packages/driver/cypress/integration/e2e/multi-domain/basic_login_spec.ts [hunk 5]
```diff
@@ -31,6 +32,7 @@ describe('basic login', { experimentalSessionSupport: true, experimentalMultiDom
     })
 
     // Scenario, Token based auth. Visit site, manually redirect to IDP hosted on secondary domain, login and redirect back to site.
+    // TODO: these were being flaky on firefox, disable until we have a trailing wait
     it('does not redirect', () => {
```
Comment (mjhenkes) [inline packages/driver/cypress/integration/e2e/multi-domain/basic_login_spec.ts:36]:
```suggestion
    it.skip('does not redirect', () => {
```

File: packages/driver/cypress/integration/e2e/multi-domain/basic_login_spec.ts [hunk 6]
```diff
@@ -53,6 +55,7 @@ describe('basic login', { experimentalSessionSupport: true, experimentalMultiDom
   // Scenario, Token based auth. Visit IDP hosted on secondary domain, login and redirect back to site.
   describe('visit secondary first', () => {
     // Problem: where does the primary domain get established
+    // TODO: these were being flaky on firefox, disable until we have a trailing wait
     it('logs in with idp redirect', () => {
```
Comment (mjhenkes) [inline packages/driver/cypress/integration/e2e/multi-domain/basic_login_spec.ts:59]:
```suggestion
    it.skip('logs in with idp redirect', () => {
```

**Review by chrisbreiding — COMMENTED** [review]
The domain argument also cannot include a path (`foobar.com/login`) or a port (`localhost:3000`). We should probably go ahead and add tests for those and maybe update the error message further.

File: packages/driver/src/cypress/error_messages.ts [hunk 7]
```diff
@@ -1694,7 +1694,7 @@ export default {
       message: `${cmd('switchToDomain')} requires enabling the experimentalMultiDomain flag`,
     },
     invalid_domain_argument: {
-      message: `${cmd('switchToDomain')} requires the first argument to be a string. You passed: \`{{arg}}\``,
+      message: `${cmd('switchToDomain')} requires the first argument to be a string representing a domain that does not not include any subdomains. You passed: \`{{arg}}\``,
```
Comment (mschile) [inline packages/driver/src/cypress/error_messages.ts:1697]: Are we able to tell them what we expected the domain to be in the error message?

Comment (mjhenkes) [reply packages/driver/src/cypress/error_messages.ts]: I've updated the error message to be more descriptive generally, but haven't tailored the message to tell them how to correct whatever string they've passed, I think that might be going too far.

File: packages/driver/cypress/integration/e2e/multi-domain/multi_domain_spec.ts [hunk 8]
```diff
@@ -145,80 +145,6 @@ describe('multi-domain', { experimentalSessionSupport: true, experimentalMultiDo
   })
 
   describe('errors', () => {
-    // @ts-ignore
```
Comment (mjhenkes) [inline packages/driver/cypress/integration/e2e/multi-domain/multi_domain_spec.ts:148]: I moved these tests to a validation specific file

File: packages/driver/package.json [hunk 9]
```diff
@@ -54,6 +54,7 @@
     "errorhandler": "1.5.1",
     "eventemitter2": "6.4.2",
     "express": "4.17.1",
+    "is-ip": "4.0.0",
```
Comment (mjhenkes) [inline packages/driver/package.json:57]: is-ip is also used by one of our dependencies, they use version 3.1.0 so now our bundle will have a duplicate. I don't think it's a big deal, but we could go back to 3.1.0 🤷‍♂️

File: packages/driver/cypress/integration/e2e/multi-domain/multi_domain_validation_specs.ts [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+    it('succeeds on a localhost domain name', () => {
+      cy.switchToDomain('localhost', () => undefined)
+    })
+
+    it('succeeds on an ip address', () => {
+      cy.switchToDomain('127.0.0.1', () => undefined)
+    })
+
+    // TODO: an ipv6 address passes validation but hangs elsewhere in the code, Come back later and fix ipv6
+    it.skip('succeeds on an ipv6 address', () => {
```
Comment (mjhenkes) [inline packages/driver/cypress/integration/e2e/multi-domain/multi_domain_validation_specs.ts:13]: Calling this out, ipv6 addresses hang for reasons outside of validation. I didn't dig in too far for fear of getting too sidetracked and missing my sprint goals

**Review by AtofStryker — APPROVED** [review]

**Review by chrisbreiding — APPROVED** [review]

**Pull request merged**